### PR TITLE
[JAX] Added prepare phase for the FusedAttnForwardFFI

### PR DESCRIPTION
**Pull request by phu0ngng**
# Description

In this PR, the FusedAttnForwardFFI is registered to XLA via 2 phases:
1. `prepare` where the `te_cudnn_handle_init_ffi` is called.
2.  `execute` phase where the actual fused attention function is called.
In the past, the default was to have only the `execute` phase.
This helps to avoid memory allocation for the handle in the cudaGraph capturing area, thus resolving the issue we have in the distributed run with cudaGraph.

In general, all the custom calls that involve cuDNN kernels should be registered with these two phases in the future. Since the prepare function needs to have the same function argument list with the execute function ones, variadic args are used to make this `te_cudnn_handle_init_ffi` general so that it can be called in any custom calls.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by phu0ngng** [comment]
@cyanguwa Please help to look at the changes in Common and let me know if there is an issue.

**Comment by phu0ngng** [comment]
\te-ci jax L1

**Comment by phu0ngng** [comment]
\te-ci jax L1

**Review by huanghua1994 — APPROVED** [review]
LGTM, let's hope this and #1314 fix the issue

File: transformer_engine/jax/cpp_extensions/custom_call.py [hunk 1]
```diff
@@ -27,8 +27,18 @@ class CustomCallAPIVersion(IntEnum):
...
-    if _name.endswith("_ffi"):
+    if _name in custom_call_with_cudnn:
+        if is_ffi_enabled():
+            jex.ffi.register_ffi_target(
+                _name,
+                {"prepare": transformer_engine_jax.te_cudnn_handle_init_ffi, "execute": _value},
+                platform="CUDA",
+                api_version=CustomCallAPIVersion.FFI.value,
+            )
+    elif _name.endswith("_ffi"):
```
[the remark this answers is not in the document]
Comment (phu0ngng) [reply transformer_engine/jax/cpp_extensions/custom_call.py]: Good idea. I implemented it and that works. Thanks.

**Review by zlsh80826 — APPROVED** [review]
LGTM

**Pull request merged**